text/0000-partial-closure-args.md

```diff
@@ -0,0 +1,357 @@
+
+// unary
+x.map(|_| 123);
+x.map(|..| 123); // same as above
+
+// binary
+x.map(|_, _| 123);
+x.map(|..| 123); // same as above
+
+// with five arguments
+x.map(|_, _, _, _, _| 123);
+x.map(|..| 123); // same as above
+```
+
+It is possible to mix underscores `_` with partial pattern matching as well with pattern matching
+arguments you want in the same way you do with tuples and structs:
+
+```rust
+// with five arguments
+x.map(|a, _, b, _, _| a + b);
+x.map(|a, _, b, ..| a + b); // same as above
+```
+
+You can also use it in the middle of the arguments list, but that requires that all the arguments
+to the right of `..` are filled until the end (without another `..` in the list):
+
+```rust
+x.map(|a, .., d, e| a + d + e);
+x.map(|a, b, .., e| a + b + e);
+```
+
+The following, however, is not allowed:
+
+```rust
+x.map(|a, .., c, .., e| a + c + e); // error: cannot have two .. in a closure argument list
+```
+
+It is important to notice that `..` is not completely isomorphic to using `_` for all arguments. If
+the arity of the function changes (for instance, it loses or gains arguments), the `..` syntax will
+still compile while the `_` will obviously fail to match, which makes using `..` in lambda
+arguments subject to breaking-changes if the type of the closure changes at the calling site.
+
+## How we teach this
+
+In the _Pattern Syntax_ section of the Rust book ([here](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#ignoring-remaining-parts-of-a-value-with-),
+some paragraphs should be added regarding pattern matching the arguments list by
+using the `..` syntax.
+
+The meaning of this syntax is equivalent to the tuple syntax:
+
+```rust
+let (a, b, ..) = (1, 2, 3, 4);
+x.map(|a, b, ..| a + b);
+```
+
+The feature is not about [variadic functions](https://en.wikipedia.org/wiki/Variadic_function): the
+arity is well-defined, so are the types of all of its arguments. Consider `..` as syntactic sugar for
+ignoring zero to several arguments.
+
+People should think of `..` as _“ignore everything else”_. So:
+
+```rust
+|a, ..|
+```
+
+is a closure that captures its first argument and ignore all the rest, whatever its arity. The
+following:
+
+```rust
+|..|
+```
+
+Is a closure that will never read any of its arguments (if any). This:
+
+```rust
+|s, .., e|
+```
+
+Is a closure that captures its first `s` argument and its last `e` argument, and ignore all the
+rest.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Interaction with type inference
+
+A closure containing a `..` – as in `|a, b, .., x|` – is not type-inferable (there is no way to
+detect which trait it implements). Closures with `..` can only be used in places where a type is
+already constrained by one of the `FnOnce`, `Fn` or `FnMut` traits. The following is, thus, not
+authorized:
+
+```rust
+let f = |a, b, ..| a + b;
+
+println!("{}", f(1, 2)); // error: cannot deduce type of f
+```
```

Comment:
I'd like to express support for this, on the basis that we already do something simi for deducing HRTB (`for<'a>`) for closures, from the `Fn` trait bounds in the "expected type", which you can observe in examples like this ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=846d46edc907480350dfe0b1303f8fa1)):

```rust
fn foo(_: impl Fn(&str)) {}

fn main() {
    foo(|_| {}); // works

    let f = |_| {};
    foo(f); // errors, cannot make the closure lifetime-polymorphic after-the-fact
}
```

cc @nikomatsakis

Reply:
Yeah, we can certainly support this as long as we're limited to cases where we have an "expected type signature".